lib/iris/fileformats/netcdf.py

```diff
@@ -1593,57 +1725,158 @@ def _get_dim_names(self, cube):
+            # get any special handling.  We *do* want to list/create them in a
+            # definite order (face,edge,node), and before non-mesh dimensions.
+            mesh_location_dimnames = {}
+            for location in MESH_LOCATIONS:
+                # Find if this location exists in the mesh, and a characteristic
+                # coordinate to identify it with.
+                # To use only _required_ UGRID components, we use a location
+                # coord for nodes, but a connectivity for faces/edges
+                if location == "node":
+                    # For nodes, identify the dim with a coordinate variable.
+                    # Selecting the X-axis one for definiteness.
+                    dim_coords = mesh.coords(include_nodes=True, axis="x")
                 else:
-                    # Return the dim_name associated with the existing
-                    # coordinate.
-                    dim_name = self._name_coord_map.name(coord)
-                    dimension_names.append(dim_name)
+                    # For face/edge, use the non-optional connectivity variable.
+                    cf_role = f"{location}_node_connectivity"
+                    dim_coords = mesh.connectivities(cf_role=cf_role)
+                if len(dim_coords) > 0:
+                    # As the mesh contains this location, we want to include this
+                    # dim in our returned mesh dims.
+                    # We should have 1 identifying variable (of either type).
+                    assert len(dim_coords) == 1
+                    mesh_coord = dim_coords[0]
+                    match = mesh_coord in self._dim_names_and_coords.coords
+                    if match:
+                        # For mesh-identifying coords, we require the *same*
+                        # coord, not an identical one (i.e. "is" not "==")
+                        name = self._dim_names_and_coords.name(mesh_coord)
+                        stored_coord = self._dim_names_and_coords.coord(name)
+                        match = mesh_coord is stored_coord
+                    if match:
+                        # Use the previous name for this dim of this mesh, from
+                        # a previously saved cube with the same mesh.
+                        dim_name = self._dim_names_and_coords.name(mesh_coord)
+                    else:
+                        if location == "node":
+                            # always 1-d
+                            (dim_length,) = mesh_coord.shape
+                        else:
+                            # extract source dim, respecting dim-ordering
+                            dim_length = mesh_coord.shape[mesh_coord.src_dim]
+                        location_dim_attr = f"{location}_dimension"
+                        dim_name = getattr(mesh, location_dim_attr)
+                        if dim_name is None:
+                            dim_name = f"{mesh.name()}_{location}"
```

Comment:
@pp-mo The use of `mesh.name()` could result in an invalid `dim_name`, right? e.g., it could be based on a space filled `long_name`

Reply:
This is tricky.
Essentially you're right, and I have added a testcase including an unsuitable long name.
Unfortunately, at least at present this code can *never* runs, because a Mesh always invents a dimension name if it didn't have one.
So for now, I've fixed this, added an assert that it never happens, _and_ a comment to explain it (!phew!)

Reply:
> a Mesh always invents a dimension name if it doesn't have one

so the line that calls `mesh.name()` can never actually run.
I have now added an assert + and explanation for this.